### ISSUE TITLE
Remove the binary block serialization step

### ISSUE THREAD

**Issue by jondubois** [LiskHQ/lisk#1621]
### Expected behavior

The process of broadcasting blocks should be simplified so that it uses as little CPU and bandwidth as possible; especially if it means simpler logic.

### Actual behavior

Right now we're serializing blocks to binary (ArrayBuffer) before broadcasting to peers. Most WebSocket implementations (including all major browsers and Node.js) support automatic `permessage-deflate` compression (https://tools.ietf.org/html/rfc7692) which can transparently compress the block at the WebSocket layer - If this is used then theoretically there shouldn't be any benefit to serializing the block to binary beforehand (entropy should be the same; and so should compressed output size). Also there is a CPU overhead cost for converting the block string to binary.

We should first check to confirm that skipping the binary serialization step doesn't lead to increased bandwidth consumption - If not, then we can remove it.

### Tasks

- [x] Remove block BSON serialization/deserialization
- [x] Transport module: postBlock test cases

**Comment by jondubois** [comment]
I did a test with a large JSON object (269K characters when stringified) then I used `netstat` to capture TCP packets:

# Without permessage-deflate:

### String:

18:11:25.868997 IP localhost.55220 > localhost.8000: Flags [P.], seq 300:314, ack 424, win 350, options [nop,nop,TS val 680150780 ecr 680149779], length 14
18:11:25.869186 IP localhost.55220 > localhost.8000: Flags [.], seq 314:22714, ack 424, win 350, options [nop,nop,TS val 680150780 ecr 680149779], length 22400
18:11:25.869210 IP localhost.8000 > localhost.55220: Flags [.], ack 22714, win 1373, options [nop,nop,TS val 680150780 ecr 680150780], length 0
18:11:25.869292 IP localhost.55220 > localhost.8000: Flags [P.], seq 22714:88197, ack 424, win 350, options [nop,nop,TS val 680150781 ecr 680150780], length 65483
18:11:25.869279 IP localhost.8000 > localhost.55220: Flags [P.], seq 424:460, ack 22714, win 1373, options [nop,nop,TS val 680150781 ecr 680150780], length 36
18:11:25.869311 IP localhost.55220 > localhost.8000: Flags [.], ack 460, win 350, options [nop,nop,TS val 680150781 ecr 680150781], length 0
18:11:25.869375 IP localhost.55220 > localhost.8000: Flags [.], seq 88197:153680, ack 460, win 350, options [nop,nop,TS val 680150781 ecr 680150781], length 65483
18:11:25.869390 IP localhost.8000 > localhost.55220: Flags [.], ack 153680, win 3420, options [nop,nop,TS val 680150781 ecr 680150781], length 0
18:11:25.869460 IP localhost.55220 > localhost.8000: Flags [.], seq 153680:219163, ack 460, win 350, options [nop,nop,TS val 680150781 ecr 680150781], length 65483
18:11:25.869514 IP localhost.55220 > localhost.8000: Flags [P.], seq 219163:269687, ack 460, win 350, options [nop,nop,TS val 680150781 ecr 680150781], length 50524
18:11:25.872724 IP localhost.8000 > localhost.55220: Flags [.], ack 269687, win 3420, options [nop,nop,TS val 680150781 ecr 680150781], length 0

**Total: 22400 + 65483 + 65483 + 65483 + 50524 = 269 KB**

### Buffer:

18:11:29.875024 IP localhost.55220 > localhost.8000: Flags [.], ack 608, win 350, options [nop,nop,TS val 680154786 ecr 680154786], length 0
18:11:30.851976 IP localhost.55220 > localhost.8000: Flags [.], seq 269695:335178, ack 608, win 350, options [nop,nop,TS val 680155763 ecr 680154786], length 65483
18:11:30.852153 IP localhost.55220 > localhost.8000: Flags [.], seq 335178:400661, ack 608, win 350, options [nop,nop,TS val 680155763 ecr 680154786], length 65483
18:11:30.852204 IP localhost.8000 > localhost.55220: Flags [.], ack 400661, win 5466, options [nop,nop,TS val 680155763 ecr 680155763], length 0
18:11:30.852573 IP localhost.55220 > localhost.8000: Flags [.], seq 400661:466144, ack 608, win 350, options [nop,nop,TS val 680155764 ecr 680155763], length 65483
18:11:30.852694 IP localhost.55220 > localhost.8000: Flags [.], seq 466144:531627, ack 608, win 350, options [nop,nop,TS val 680155764 ecr 680155763], length 65483
18:11:30.852725 IP localhost.55220 > localhost.8000: Flags [P.], seq 531627:539082, ack 608, win 350, options [nop,nop,TS val 680155764 ecr 680155763], length 7455
18:11:30.852763 IP localhost.8000 > localhost.55220: Flags [.], ack 531627, win 7512, options [nop,nop,TS val 680155764 ecr 680155764], length 0
18:11:30.876358 IP localhost.8000 > localhost.55220: Flags [P.], seq 608:644, ack 539082, win 8536, options [nop,nop,TS val 680155788 ecr 680155764], length 36

**Total: 65483 + 65483 + 65483 + 65483 + 7455 = 269 KB**

----

# With permessage-deflate:

### String:

18:27:27.187810 IP localhost.55250 > localhost.8000: Flags [.], ack 534, win 350, options [nop,nop,TS val 681112095 ecr 681112095], length 0
18:27:27.199566 IP localhost.55250 > localhost.8000: Flags [P.], seq 57243:57251, ack 534, win 350, options [nop,nop,TS val 681112107 ecr 681112095], length 8
18:27:27.199703 IP localhost.55250 > localhost.8000: Flags [P.], seq 57251:114186, ack 534, win 350, options [nop,nop,TS val 681112107 ecr 681112095], length 56935
18:27:27.199740 IP localhost.8000 > localhost.55250: Flags [.], ack 114186, win 3420, options [nop,nop,TS val 681112107 ecr 681112107], length 0
18:27:28.188962 IP localhost.8000 > localhost.55250: Flags [P.], seq 534:570, ack 114186, win 3420, options [nop,nop,TS val 681113096 ecr 681112107], length 36

**Total: 56935 = 57 KB**

### Buffer:

18:27:31.193193 IP localhost.55250 > localhost.8000: Flags [.], ack 682, win 350, options [nop,nop,TS val 681116100 ecr 681116100], length 0
18:27:32.190059 IP localhost.55250 > localhost.8000: Flags [P.], seq 114194:114202, ack 682, win 350, options [nop,nop,TS val 681117097 ecr 681116100], length 8
18:27:32.190195 IP localhost.55250 > localhost.8000: Flags [P.], seq 114202:171137, ack 682, win 350, options [nop,nop,TS val 681117097 ecr 681116100], length 56935
18:27:32.190220 IP localhost.8000 > localhost.55250: Flags [.], ack 171137, win 4443, options [nop,nop,TS val 681117097 ecr 681117097], length 0

**Total: 56935 = 57 KB**

^ Ignore the very small packets because they're just heartbeats and are not relevant to this test.

### Conclusions:

- It doesn't appear to make any difference if the data is serialized to a Buffer first or sent directly as a string (regardless of permessage-deflate).
- For large payloads, using permessage-deflate compresses the data to almost 1/5th of its original size :)
- permessage-deflate is on by default in SocketCluster

### Caveat:

The `uws` WebSocket engine which SocketCluster uses by default currently only supports unidirectional permessage-deflate (so for client -> server messages but not for server -> client messages). This should be sufficient for our use case though since RPC responses are usually tiny and don't need compression.

**Comment by jondubois** [comment]
Actually I made a small mistake in the previous test. I wasn't creating the buffer using `bson.serialize()` (which is the serialization function that we're currently using). It seems that serialising to BSON actually makes the buffer slightly larger (which I did not expect) both with and without permessage-deflate:

## Without permessage-deflate:

### BSON Buffer

```bash
14:17:53.545637 IP localhost.42110 > localhost.8000: Flags [.], seq 539082:604565, ack 788, win 350, options [nop,nop,TS val 708770082 ecr 708769124], length 65483
14:17:53.545689 IP localhost.42110 > localhost.8000: Flags [.], seq 604565:670048, ack 788, win 350, options [nop,nop,TS val 708770082 ecr 708769124], length 65483
14:17:53.545703 IP localhost.8000 > localhost.42110: Flags [.], ack 670048, win 10799, options [nop,nop,TS val 708770082 ecr 708770082], length 0
14:17:53.545716 IP localhost.42110 > localhost.8000: Flags [.], seq 670048:735531, ack 788, win 350, options [nop,nop,TS val 708770082 ecr 708769124], length 65483
14:17:53.545747 IP localhost.42110 > localhost.8000: Flags [.], seq 735531:801014, ack 788, win 350, options [nop,nop,TS val 708770082 ecr 708770082], length 65483
14:17:53.545752 IP localhost.8000 > localhost.42110: Flags [.], ack 801014, win 12846, options [nop,nop,TS val 708770082 ecr 708770082], length 0
14:17:53.545782 IP localhost.42110 > localhost.8000: Flags [P.], seq 801014:856079, ack 788, win 350, options [nop,nop,TS val 708770082 ecr 708770082], length 55065
```

Total: 65483 + 65483 + 65483 + 65483 + 55065 = **317 KB**

----

## With permessage-deflate:

### BSON Buffer

```bash
14:20:18.323319 IP localhost.42116 > localhost.8000: Flags [.], ack 862, win 350, options [nop,nop,TS val 708914859 ecr 708914859], length 0
14:20:19.289059 IP localhost.42116 > localhost.8000: Flags [P.], seq 114194:114208, ack 862, win 350, options [nop,nop,TS val 708915825 ecr 708914859], length 14
14:20:19.289126 IP localhost.42116 > localhost.8000: Flags [.], seq 114208:179691, ack 862, win 350, options [nop,nop,TS val 708915825 ecr 708914859], length 65483
14:20:19.289135 IP localhost.8000 > localhost.42116: Flags [.], ack 179691, win 4443, options [nop,nop,TS val 708915825 ecr 708915825], length 0
14:20:19.289149 IP localhost.42116 > localhost.8000: Flags [P.], seq 179691:185463, ack 862, win 350, options [nop,nop,TS val 708915825 ecr 708915825], length 5772
14:20:19.322835 IP localhost.8000 > localhost.42116: Flags [P.], seq 862:898, ack 185463, win 5466, options [nop,nop,TS val 708915859 ecr 708915825], length 36
14:20:19.322859 IP localhost.42116 > localhost.8000: Flags [.], ack 898, win 350, options [nop,nop,TS val 708915859 ecr 708915859], length 0
```

Total: 65483 + 5772 = **71 KB**

Anyway it doesn't change the conclusion of the test.

**Comment by jondubois** [comment]
I've implemented this fix but I haven't written the unit test cases for it yet. It would be nice to get https://github.com/LiskHQ/lisk/pull/1644 merged in first.

**Comment by jondubois** [comment]
Note: When https://github.com/LiskHQ/lisk/pull/1619 is merged into 1.0.0, merge changes into this branch and modify the schema because it won't be receiving `{base64: true, data:'98238237kjshsf=='}` objects anymore.
